aws/resource_aws_cur_report_definition.go

```diff
@@ -81,34 +84,129 @@ func resourceAwsCurReportDefinition() *schema.Resource {
+					false,
+				),
+			},
 		},
 	}
 }
 
 func resourceAwsCurReportDefinitionCreate(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).costandusagereportconn
 
+	additionalArtifacts := expandStringSet(d.Get("additional_artifacts").(*schema.Set))
+	compression := aws.String(d.Get("compression").(string))
+	format := aws.String(d.Get("format").(string))
+	prefix := aws.String(d.Get("s3_prefix").(string))
+	reportVersioning := aws.String(d.Get("report_versioning").(string))
+
+	hasAthena := false
+
+	// perform various combination checks, AWS API unhelpfully just returns an empty ValidationException
+	// these combinations have been determined from the Create Report AWS Console Web Form
+	for _, artifact := range additionalArtifacts {
+		if *artifact == costandusagereportservice.AdditionalArtifactAthena {
+			hasAthena = true
+			break
+		}
+	}
+
+	if hasAthena {
+		if len(additionalArtifacts) > 1 {
+			return fmt.Errorf(
+				"When %s exists within additional_artifacts, no other artifact type can be declared",
+				costandusagereportservice.AdditionalArtifactAthena,
+			)
+		}
+
+		if len(*prefix) == 0 {
+			return fmt.Errorf(
+				"When %s exists within additional_artifacts, prefix cannot be empty",
+				costandusagereportservice.AdditionalArtifactAthena,
+			)
+		}
+
+		if *reportVersioning != costandusagereportservice.ReportVersioningOverwriteReport {
+			return fmt.Errorf(
+				"When %s exists within additional_artifacts, report_versioning must be %s",
+				costandusagereportservice.AdditionalArtifactAthena,
+				costandusagereportservice.ReportVersioningOverwriteReport,
+			)
+		}
+	}
+
+	if *format == costandusagereportservice.ReportFormatParquet {
+		if *compression != costandusagereportservice.CompressionFormatParquet {
+			return fmt.Errorf(
+				"When format is %s, compression must also be %s",
+				costandusagereportservice.ReportFormatParquet,
+				costandusagereportservice.CompressionFormatParquet,
+			)
+		}
+	} else {
+		if *compression == costandusagereportservice.CompressionFormatParquet {
+			return fmt.Errorf(
+				"When format is %s, format must not be %s",
+				*format,
+				costandusagereportservice.CompressionFormatParquet,
+			)
+		}
+	}
+
+	if hasAthena && (*format != costandusagereportservice.ReportFormatParquet) {
+		return fmt.Errorf(
+			"When %s exists within additional_artifacts, both format and compression must be %s",
+			costandusagereportservice.AdditionalArtifactAthena,
+			costandusagereportservice.ReportFormatParquet,
+		)
+	}
+
+	if !hasAthena && len(additionalArtifacts) > 1 && (*format == costandusagereportservice.ReportFormatParquet) {
+		return fmt.Errorf(
+			"When additional_artifacts includes %s and/or %s, format must not be %s",
+			costandusagereportservice.AdditionalArtifactQuicksight,
+			costandusagereportservice.AdditionalArtifactRedshift,
+			costandusagereportservice.ReportFormatParquet,
+		)
+	}
+	// end checks
```

Comment:
Minor suggestion to move this logic into a separate function that can the be unit tested.

Reply:
@ewbankkit makes sense